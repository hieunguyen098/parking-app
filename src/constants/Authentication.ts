export enum AuthMethod {
    CHECK_PHONE = 'check-phone',
    LOGIN = 'login',
    SIGNUP = 'sign-up',
    VERIFY_PHONE = 'verify-phone',
    FORGOT_PASSWORD = 'set-new-password',
}
