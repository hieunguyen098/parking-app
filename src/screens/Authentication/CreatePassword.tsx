import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import LargeButton from '../../components/Buttons/LargeButton';
import { useNavigation } from '@react-navigation/native';
import PasswordInput from './components/PasswordInput/PasswordInput';
import PinCodeInput from './components/PinCodeInput/PinCodeInput';

const CreatePassword = () => {
    const navigation: any = useNavigation();
    const nextStep = () => {
        navigation.navigate('CreatePassword');
    };
    return (
        <View style={styles.container}>
            <View>
                <Text style={styles.description}>Nhập mật khẩu của bạn</Text>
                <PasswordInput length={6} />
                <Text style={styles.description}>Nhập lại mật khẩu</Text>
                <PasswordInput length={6} />
                <PinCodeInput length={4} />
            </View>
            <LargeButton onPress={nextStep} type="primary" title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default CreatePassword;

const styles = StyleSheet.create({
    container: {
        backgroundColor: '#fff',
        padding: 20,
        flex: 1,
        justifyContent: 'space-between',
    },
    description: {
        fontSize: 16,
    },
    continueButton: {
        marginBottom: 24,
        borderRadius: 100,
    },
});
