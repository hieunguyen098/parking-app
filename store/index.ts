import { configureStore } from '@reduxjs/toolkit';
import { authReducer } from './slices/authSlice';
import { generalReducer } from './slices/generalSlice';

export const store = configureStore({
    reducer: {
        auth: authReducer,
        general: generalReducer,
    },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
