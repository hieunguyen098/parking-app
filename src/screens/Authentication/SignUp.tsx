import { StyleSheet, View, Image, Pressable } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import LargeButton from '../../components/Buttons/LargeButton';
import FieldInput from './components/FieldInput';
import DateInput from './components/DateInput';
import Select from './components/Select';
import * as ImagePicker from 'expo-image-picker';
import { GlobalStyles } from '../../constants/style';

const genders = [
    {
        id: 0,
        label: 'Nam',
        value: 1,
    },
    {
        id: 1,
        label: 'Nữ',
        value: 0,
    },
    {
        id: 2,
        label: 'Khác',
        value: -1,
    },
];
const SignUp = () => {
    const navigation: any = useNavigation();
    const nextStep = () => {
        navigation.navigate('CreatePassword');
    };
    const [image, setImage] = useState('');

    const pickImage = async () => {
        // No permissions request is necessary for launching the image library
        let result = await ImagePicker.launchImageLibraryAsync({
            mediaTypes: ImagePicker.MediaTypeOptions.Images,
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        if (!result.canceled) {
            setImage(result.assets[0].uri);
        }
    };

    const takePhoto = async () => {
        let result = await ImagePicker.launchCameraAsync({
            allowsEditing: true,
            aspect: [4, 4],
            quality: 1,
        });

        if (!result.canceled) {
            setImage(result.assets[0].uri);
        }
    };

    return (
        <View style={styles.container}>
            <View>
                <Pressable onPress={takePhoto}>
                    <Image
                        source={image ? { uri: image } : require('../../../assets/images/default-avt-0.png')}
                        style={styles.avatar}
                    />
                </Pressable>
                <Pressable style={styles.editAvatarBtn} onPress={pickImage}>
                    <Image
                        style={{ width: '100%', height: '100%' }}
                        source={require('../../../assets/images/edit-icon.png')}
                    />
                </Pressable>
            </View>
            <View style={styles.form}>
                <FieldInput placeHolder="Họ tên" iconPosition="none" keyboardType="text" />
                <DateInput
                    placeHolder="Ngày sinh"
                    source={require('../../../assets/images/calendar-icon.png')}
                    iconPosition="right"
                    keyboardType="text"
                />
                <FieldInput
                    placeHolder="Email"
                    source={require('../../../assets/images/mail-icon.png')}
                    iconPosition="right"
                    keyboardType="email-address"
                />
                <Select data={genders} />
            </View>
            <LargeButton onPress={nextStep} type="primary" title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default SignUp;

const styles = StyleSheet.create({
    container: {
        backgroundColor: '#fff',
        padding: 20,
        flex: 1,
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    continueButton: {
        marginBottom: 24,
        borderRadius: 100,
    },
    avatar: {
        width: 200,
        height: 200,
        borderRadius: 100,
    },
    form: {
        width: '100%',
    },
    editAvatarBtn: {
        position: 'absolute',
        right: 0,
        bottom: 0,
        backgroundColor: GlobalStyles.colors.primaryOrange,
        padding: 5,
        width: 40,
        height: 40,
        borderRadius: 5,
    },
});
