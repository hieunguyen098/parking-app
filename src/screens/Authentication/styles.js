import { StyleSheet } from 'react-native';
import { GlobalStyles } from '../../constants/style';
const styles = StyleSheet.create({
    container: {
        backgroundColor: '#fff',
        padding: 20,
        flex: 1,
        justifyContent: 'space-between',
        alignItems: 'center',
    },
    description: {
        fontSize: 16,
    },
    continueButton: {
        marginBottom: 24,
        borderRadius: 100,
    },
    avatar: {
        width: 150,
        height: 150,
        borderRadius: 100,
    },
    form: {
        width: '100%',
    },
    editAvatarBtn: {
        position: 'absolute',
        right: 0,
        bottom: 0,
        backgroundColor: GlobalStyles.colors.primaryOrange,
        padding: 5,
        width: 30,
        height: 30,
        borderRadius: 5,
    },
    group: {
        width: '100%',
    },
    forgotPasswordBtn: {
        fontSize: 16,
        color: GlobalStyles.colors.primaryOrange,
        textAlign: 'center',
        marginTop: 10,
    },
    textCenter: {
        fontSize: 16,
        textAlign: 'center',
    },
    bottomContainer: {
        flexDirection: 'row',
        justifyContent: 'space-between',
    },
    fingerprintContainer: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'center',
        marginTop: 16,
    },
    fingerprintBtn: { fontSize: 16, paddingLeft: 8 },
});

export default styles;
