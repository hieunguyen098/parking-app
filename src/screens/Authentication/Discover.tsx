import {
    ActivityIndicator,
    Alert,
    Keyboard,
    KeyboardAvoidingView,
    Modal,
    Text,
    TouchableOpacity,
    TouchableWithoutFeedback,
    View,
} from 'react-native';
import React, { useEffect, useState } from 'react';
import LargeButton from '../../components/Buttons/LargeButton';
import { useFocusEffect, useNavigation } from '@react-navigation/native';
import styles from './styles';
import AsyncStorage from '@react-native-async-storage/async-storage';
import FieldInput from '../../components/FieldInput';
import { useDispatch, useSelector } from 'react-redux';
import { authActions } from '../../../store/slices/authSlice';
import { validatePhoneNumber } from '../../utils/validation';
import { discover } from '../../services/auth';
import { SafeAreaView } from 'react-native-safe-area-context';
import * as SecureStore from 'expo-secure-store';
import { getAccount, removeAccount } from '../../utils/account.helper';

const Discover = () => {
    const [phone, setPhone] = useState('');
    const user = useSelector((state: any) => state.auth.user);
    const dispatch = useDispatch();
    const navigation: any = useNavigation();
    const [invalid, setInvalid] = useState(false);
    const [isLoading, setIsLoading] = useState(false);

    const clearAll = async () => {
        try {
            await AsyncStorage.clear();
        } catch (e) {
            // clear error
        }

        console.log('Done.');
    };

    useFocusEffect(
        React.useCallback(() => {
            (async () =>
                await AsyncStorage.getItem('account')
                    .then((value) => {
                        if (value !== null) {
                            setPhone(JSON.parse(value).phoneNumber);
                        } else {
                            console.log('Key not found!');
                        }
                    })
                    .catch((error) => console.log(`Error retrieving data: ${error}`)))();
        }, []),
    );

    const startDiscovering = async () => {
        setIsLoading(true);

        if (validatePhoneNumber(phone)) {
            dispatch(authActions.setUser({ phone }));
            const account = await getAccount();
            if (account === null || account.phoneNumber !== phone) {
                await removeAccount();
                await SecureStore.deleteItemAsync('password');
            }
            const data = await discover(phone);

            if (data.data && data.data.length > 0) {
                dispatch(authActions.setUser(data.data[0]));
                if (account === null || account.phoneNumber !== phone) {
                    const jsonValue = JSON.stringify({ phoneNumber: data.data[0]?.phoneNumber });
                    await AsyncStorage.setItem('account', jsonValue)
                        .then(() => console.log('Data saved successfully!'))
                        .catch((error) => console.log(`Error saving data: ${error}`));
                }
                navigation.navigate('Login');
            } else {
                dispatch(authActions.setUser({ phone }));
                navigation.navigate('SignUp');
            }
        } else {
            setInvalid(true);
        }
        setIsLoading(false);
    };

    const clearViewedOnboarding = async () => {
        await AsyncStorage.removeItem('@viewedOnboarding');
    };

    return (
        <TouchableWithoutFeedback style={styles.group} onPress={Keyboard.dismiss}>
            <View style={styles.container}>
                <Modal animationType="fade" transparent={true} visible={isLoading}>
                    <View
                        style={{
                            flex: 1,
                            backgroundColor: 'rgba(0, 0, 0, 0.5)',
                            justifyContent: 'center',
                            alignItems: 'center',
                        }}
                    >
                        <ActivityIndicator size="large" color="#fff" />
                    </View>
                </Modal>
                <View style={styles.group}>
                    <Text style={styles.description}>Nhập số điện thoại để đăng ký hoặc đăng nhập</Text>
                    <FieldInput
                        setValue={setPhone}
                        autoFocus={true}
                        value={phone}
                        placeHolder="Số điện thoại"
                        source={require('../../../assets/images/phone-icon.png')}
                        iconPosition="left"
                        keyboardType="phone-pad"
                        warning={{
                            show: invalid,
                            message: 'Số điện thoại không hợp lệ!',
                        }}
                        onChange={() => setInvalid(false)}
                        maxLength={10}
                    />
                    <LargeButton
                        onPress={() => navigation.navigate('TabBarScreen')}
                        title="Vào trang chủ"
                        style={[styles.continueButton, { backgroundColor: 'lightblue' }]}
                    />
                    <LargeButton
                        onPress={clearViewedOnboarding}
                        title="clear ViewedOnboarding"
                        type="secondary"
                        style={styles.continueButton}
                    />
                </View>
                <LargeButton onPress={startDiscovering} type="primary" title="Tiếp tục" style={styles.continueButton} />
            </View>
        </TouchableWithoutFeedback>
    );
};

export default Discover;
