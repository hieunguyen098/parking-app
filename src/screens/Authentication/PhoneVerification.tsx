import { StyleSheet, Text, View } from 'react-native';
import React, { useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import LargeButton from '../../components/Buttons/LargeButton';
import styles from './styles';
import PinCodeInput from '../../components/PinCodeInput/PinCodeInput';
import { verifyPhoneNumber } from '../../services/auth';
import { useSelector } from 'react-redux';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { useCountDown } from '../../hooks';

const PhoneVerification = ({ navigation }: any) => {
    const [otp, setOtp] = useState('');
    const [invalid, setInvalid] = useState(false);
    const user = useSelector((state: any) => state.auth.user);
    const { remainingTime, setRemainingTime } = useCountDown();

    const nextStep = async () => {
        const data = await verifyPhoneNumber(user.phone, otp);
        console.log(data);
        if (data.returnCode > 0) {
            if (!data.data) {
                return;
            }
            AsyncStorage.setItem('accessToken', data.data[0].accessToken);
            navigation.navigate('TabBarScreen');
        } else {
            setInvalid(true);
        }
    };
    return (
        <View style={styles.container}>
            <View style={styles.group}>
                <Text style={styles.description}>
                    Mã xác thực được gửi đến {user.phone.slice(0, 3)}****{user.phone.slice(7, 10)}
                </Text>

                <PinCodeInput
                    warning={{ show: invalid, message: 'Mã OTP không hợp lệ.' }}
                    value={otp}
                    setValue={setOtp}
                    length={6}
                />
                <Text style={styles.textCenter}>Gửi lại mã sau {remainingTime}s</Text>
            </View>
            <LargeButton onPress={nextStep} title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default PhoneVerification;
