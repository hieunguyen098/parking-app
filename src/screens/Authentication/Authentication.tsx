import { StyleSheet, Text, View } from 'react-native';
import React from 'react';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import Login from './Login/Login';
import Discover from './Discover';
import SignUp from './SignUp/SignUp';
import CreatePassword from './SignUp/CreatePassword';
import ForgotPassword from './ForgotPassword/ForgotPassword';
import PhoneVerification from './PhoneVerification';
import ResetPassword from './ForgotPassword/ResetPassword';

const Stack = createNativeStackNavigator();

const Authentication = () => {
    return (
        <Stack.Navigator
            screenOptions={{
                statusBarColor: '#fff',
            }}
        >
            <Stack.Screen
                name="Discover"
                component={Discover}
                options={{
                    title: 'Khám phá ngay',
                    presentation: 'modal',
                }}
            />
            <Stack.Screen
                name="Login"
                component={Login}
                options={{
                    title: 'Đăng nhập',
                    presentation: 'modal',
                }}
            />
            <Stack.Screen
                name="SignUp"
                component={SignUp}
                options={{
                    title: 'Tạo tài khoản',
                    presentation: 'modal',
                }}
            />
            <Stack.Screen
                name="CreatePassword"
                component={CreatePassword}
                options={{
                    title: 'Thiết lập mật khẩu',
                    presentation: 'modal',
                }}
            />
            <Stack.Screen
                name="PhoneVerification"
                component={PhoneVerification}
                options={{
                    title: 'Xác thực số điện thoại',
                    presentation: 'modal',
                }}
            />
            <Stack.Screen
                name="ForgotPassword"
                component={ForgotPassword}
                options={{
                    title: 'Quên mật khẩu',
                    presentation: 'modal',
                }}
            />
            <Stack.Screen
                name="ResetPassword"
                component={ResetPassword}
                options={{
                    title: 'Thiết lập mật khẩu mới',
                    presentation: 'modal',
                }}
            />
        </Stack.Navigator>
    );
};

export default Authentication;

const styles = StyleSheet.create({});
