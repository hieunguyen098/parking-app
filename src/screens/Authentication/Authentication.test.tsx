import React from 'react';
import { render } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';

import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { store } from '../../../store';
import Authentication from './Authentication';

jest.useFakeTimers();
test('test settings', () => {
    const queryClient = new QueryClient();
    const component = (
        <QueryClientProvider client={queryClient}>
            <Provider store={store}>
                <NavigationContainer>
                    <Authentication />
                </NavigationContainer>
            </Provider>
        </QueryClientProvider>
    );

    const tree = render(component);
    expect(tree).toMatchSnapshot();
});
