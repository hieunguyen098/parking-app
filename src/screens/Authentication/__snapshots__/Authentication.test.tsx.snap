// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test settings 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      customAnimationOnSwipe={true}
      forwardedRef={[Function]}
      fullScreenSwipeEnabled={true}
      gestureResponseDistance={
        Object {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      stackAnimation="slide_from_bottom"
      stackPresentation="push"
      statusBarColor="#fff"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="vertical"
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Khám phá ngay"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          Object {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
              },
              Object {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "flex": 1,
                "justifyContent": "space-between",
                "padding": 20,
              }
            }
          >
            <Modal
              animationType="fade"
              hardwareAccelerated={false}
              transparent={true}
              visible={false}
            />
            <View
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "fontSize": 16,
                    "fontWeight": "bold",
                  }
                }
              >
                Nhập số điện thoại để đăng ký hoặc đăng nhập
              </Text>
              <View>
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "#FBFBFB",
                        "borderColor": "transparent",
                        "borderRadius": 20,
                        "borderWidth": 1,
                        "flexDirection": "row",
                        "marginVertical": 10,
                        "paddingHorizontal": 20,
                      },
                      false,
                    ]
                  }
                >
                  <Image
                    source={1}
                    style={
                      Array [
                        Object {
                          "height": 25,
                          "width": 25,
                        },
                        Object {
                          "marginRight": 10,
                        },
                      ]
                    }
                  />
                  <TextInput
                    autoFocus={true}
                    keyboardType="phone-pad"
                    maxLength={10}
                    onChange={[Function]}
                    onChangeText={[Function]}
                    placeholder="Số điện thoại"
                    style={
                      Object {
                        "flex": 1,
                        "fontSize": 16,
                        "paddingVertical": 20,
                      }
                    }
                    value=""
                  />
                </View>
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "#FF953A",
                    "borderRadius": 8,
                    "paddingHorizontal": 12,
                    "paddingVertical": 18,
                    "width": "100%",
                  },
                  Object {
                    "backgroundColor": "#FF953A",
                  },
                  Object {
                    "borderRadius": 100,
                    "marginBottom": 24,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "fontSize": 18,
                      "fontWeight": "600",
                      "textAlign": "center",
                    },
                    Object {
                      "color": "#fff",
                    },
                    undefined,
                  ]
                }
              >
                Tiếp tục
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
